src/main/java/com/mathworks/ci/actions/RunMatlabCommandAction.java

```diff
@@ -7,38 +7,62 @@
 
 import java.io.IOException;
 
+import com.mathworks.ci.BuildConsoleAnnotator;
+import com.mathworks.ci.MatlabBuilderConstants;
 import com.mathworks.ci.MatlabExecutionException;
-import com.mathworks.ci.parameters.RunActionParameters;
+import com.mathworks.ci.parameters.CommandActionParameters;
 import com.mathworks.ci.utilities.MatlabCommandRunner;
 
-public class RunMatlabCommandAction {
-    private RunActionParameters params; 
-    private MatlabCommandRunner runner;
+import hudson.model.Run;
 
-    public RunMatlabCommandAction(MatlabCommandRunner runner, RunActionParameters params) {
-        this.runner = runner;
+public class RunMatlabCommandAction extends MatlabAction {
+    private CommandActionParameters params;
+
+    public RunMatlabCommandAction(MatlabCommandRunner runner, BuildConsoleAnnotator annotator, CommandActionParameters params) {
+        super(runner, annotator);
         this.params = params;
     }
 
-    public RunMatlabCommandAction(RunActionParameters params) throws IOException, InterruptedException {
-        this(new MatlabCommandRunner(params), params);
+    public RunMatlabCommandAction(CommandActionParameters params) throws IOException, InterruptedException {
+        this(new MatlabCommandRunner(params), 
+                new BuildConsoleAnnotator(
+                    params.getTaskListener().getLogger(), 
+                    params.getBuild().getCharset()),
+                params);
     }
 
     public void run() throws IOException, InterruptedException, MatlabExecutionException {
+        super.copyBuildPluginsToTemp();
+        super.setBuildEnvVars();
+
+        // Redirect output to the build annotator
+        runner.redirectStdOut(annotator);
+
+        // Prepare MATLAB command
+        String command = "addpath('" 
+            + runner.getTempFolder().getRemote()
+            + "'); " + this.params.getCommand();
+
         try {
-            runner.runMatlabCommand(this.params.getCommand());
+            runner.runMatlabCommand(command);
         } catch (Exception e) {
             this.params.getTaskListener().getLogger()
                 .println(e.getMessage());
             throw(e);
         } finally {
-            try {
-                this.runner.removeTempFolder();
-            } catch (Exception e) {
-                // Don't want to override more important error
-                // thrown in catch block
-                System.err.println(e.toString());
-            }
+            annotator.forceEol();
+
+            teardownAction();
         }
     }
+
+    @Override
+    public void teardownAction() {
+        Run<?, ?> build = this.params.getBuild();
+
+        // Handle build result
+        super.moveJsonArtifactToBuildRoot(build, MatlabBuilderConstants.BUILD_ARTIFACT);
+        
+        super.teardownAction();
+    }
```

Comment:
This is kind of an example of the code duplication that I mentioned in my last comment that could get a little bit painful to maintain over time. Again definitely not a blocker.